Automattic/wp-calypso
# composite-checkout: Use constants for transaction and form statuses

#### Changes proposed in this Pull Request

* Create an enum for statuses
* Uses a base interface with commonalities between payloads—the reducer maps these fields to state with nulls where values are empty
* Leverage this base interface for each payload type, using `Pick` and `Required` to focus the type—with unique a `type` property
* Create a payload union type for each payload type to allow type guarding
* For each dispatching function return, pass its type with the `useCallback` generic to type incoming arguments

#### Testing instructions



Other than the type themselves, there is a change in the transaction status handler which shouldn't affect logic.

The form status validator now runs against the enum and should be testable each status change.

It's worth running through some of the common testing scenarios, although for the most part the unit and e2e tests should serve to prevent regressions.

Fixes #44999 #45333
